### PR TITLE
feature: Allow for multiple public keys for verifying JWTs

### PR DESCRIPTION
**Pull request by Don-CA**
## Purpose

Allow for multiple public keys for verifying JWTs

## Context

Currently we can only verify JWTs signed by a single private key. Almost always that is from web-gateway. However, JWTs signed by other services need to be supported. With this change, we can pass in the a map of `kid`->`PUBLIC KEY`. 

Consumers will then be easily able to add more keys much like the [ruby version](https://github.com/cultureamp/ca-authentication-ruby/blob/ceb97944619ae9d6cb138cb47cd7279150d67412/README.md?plain=1#L103) of this functionality.

There are no changes to the existing API for constructing the Decoder.

This unblocks [perform core](https://cultureamp.atlassian.net/jira/software/c/projects/SR2/issues/SR2-646) 

## Notes

Quite a bit of cleanup was required here, package refreshes and action tweaks.

### CODE REVIEW

File: .github/workflows/pipeline.yml [hunk 1]
```diff
@@ -2,45 +2,42 @@ name: pipeline
...
-        # https://securego.io/docs/rules/rule-intro.html
-        run: |
-          export PATH=$PATH:$(go env GOPATH)/bin
-          go get -u github.com/securego/gosec/v2/cmd/gosec@v2.8.0
-          gosec -fmt json -out gosec-results.json -conf gosec-config.json -stdout ./...
+        uses: securego/gosec@master
+        with:
+          args: -fmt json -out gosec-results.json -conf gosec-config.json -stdout ./...
   
       - uses: actions/upload-artifact@v2
```
Comment (techmanmike) [inline .github/workflows/pipeline.yml:42]: should this be v3?

File: .golangci.yml [hunk 2]
```diff
@@ -7,8 +7,6 @@ linters:
   disable-all: true
   enable:
     - bodyclose
-    - deadcode
```
Comment (techmanmike) [inline .golangci.yml:10]: Why did you remove "deadcode"?

File: jwt/jwtDecoder.go [hunk 3]
```diff
@@ -30,9 +31,23 @@ func NewDecoderFromPath(pubKeyPath string) (Decoder, error) {
...
+}
+
+// NewDecoderFromBytes creates a new Decoder given the public key as a []byte
+func NewMultiKeyDecoderFromBytes(verifyBytes []byte, additionalVerifyBytes map[string][]byte) (Decoder, error) {
 	verifyKey, err := jwtgo.ParseRSAPublicKeyFromPEM(verifyBytes)
+	additionalKeys := make(map[string]*rsa.PublicKey)
+	for key, bytes := range additionalVerifyBytes {
+		additionalKey, err := jwtgo.ParseRSAPublicKeyFromPEM(bytes)
+		if err != nil {
+			return Decoder{}, err
```
Comment (techmanmike) [inline jwt/jwtDecoder.go:44]: If there are multiple additionals, but the first one is invalid (won't parse) then we will NOT process the remaining additionals... Should we change this so that we will process all the additions and only add the valid ones?

**Review by techmanmike — DISMISSED** [review]

File: jwt/jwtDecoder.go (hunk 3 above)
Comment (Don-CA) [reply jwt/jwtDecoder.go:44]: no, we should blow up on invalid configuration and fail the start up of the container. This will fail the deployment of new containers and won't shut down the current ones, minimising any risk of outage.

File: .golangci.yml (hunk 2 above)
Comment (Don-CA) [reply .golangci.yml:10]: a bunch of linters are deprecated/abandoned
https://golangci-lint.run/usage/linters/

In the case of deadcode: The owner seems to have abandoned the linter. Replaced by unused.

File: authz/client.go [hunk 4]
```diff
@@ -65,6 +65,7 @@ func (client Client) EvaluateBooleanPolicy(ctx context.Context, policy string, i
 	if err != nil {
 		return nil, err // if there is a compile error, etc. assume the kill switch is OFF
 	}
+	defer response.Body.Close()
```
Comment (Don-CA) [inline authz/client.go:68]: linter was complaining that a body wasn't getting closed

**Review by techmanmike — APPROVED** [review]
